crates/ra_hir/src/code_model_api.rs

```diff
@@ -0,0 +1,26 @@
+use ra_db::{CrateId, Cancelable};
+
+use crate::{Module, Name, db::HirDatabase};
+
+/// hir::Crate describes a single crate. It's the main inteface with which
+/// crate's dependencies interact. Mostly, it should be just a proxy for the
+/// root module.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Crate {
+    pub(crate) crate_id: CrateId,
+}
+
+#[derive(Debug)]
+pub struct CrateDependency {
+    pub krate: Crate,
+    pub name: Name,
+}
+
+impl Crate {
+    pub fn dependencies(&self, db: &impl HirDatabase) -> Vec<CrateDependency> {
+        self.dependencies_impl(db)
+    }
+    pub fn root_module(&self, db: &impl HirDatabase) -> Cancelable<Option<Module>> {
```

Comment:
Looking at it from the outside, it seems kind of random that this returns `Cancelable`, but `dependencies` doesn't...

Reply:
Yeah, I guess everything should just be `Cancelable`.


I am also debating implement cancellation via unwinding. It will certainly reduce the API ceremony quite a bit. It may be argued that this is not what panicing was made for, but most other systems handle cancellation via unwinding as well...